pinot-core/src/main/java/org/apache/pinot/core/query/selection/SelectionOperatorService.java

```diff
@@ -96,28 +146,114 @@ public void reduceWithOrdering(Collection<DataTable> dataTables) {
               row[colId] = null;
             }
           }
-          SelectionOperatorUtils.addToPriorityQueue(row, _rows, _numRowsToKeep);
+          rightRows.add(row);
           Tracing.ThreadAccountantOps.sampleAndCheckInterruptionPeriodically(rowId);
         }
       } else {
         for (int rowId = 0; rowId < numRows; rowId++) {
           Object[] row = SelectionOperatorUtils.extractRowFromDataTable(dataTable, rowId);
-          SelectionOperatorUtils.addToPriorityQueue(row, _rows, _numRowsToKeep);
+          rightRows.add(row);
           Tracing.ThreadAccountantOps.sampleAndCheckInterruptionPeriodically(rowId);
         }
       }
+      dataTablesRowList.add(rightRows);
     }
+    return dataTablesRowList;
   }
 
   /**
-   * Renders the selection rows to a {@link ResultTable} object for selection queries with <code>ORDER BY</code>.
+   * Merge sorted dataTables using N-way merge
+   * @param dataTables sorted dataTables
+   * @return sorted rows
    */
-  public ResultTable renderResultTableWithOrdering() {
+  private List<Object[]> nWayMergeDataTables(Collection<DataTable> dataTables) {
+    List<List<Object[]>> dataTablesRowList = processDataTableRows(dataTables);
```

Comment:
We don't want to process all rows up-front because most of them might just be thrown away. We can process them as we add rows into the heap